lib/wallaroo/core/sink/connector_sink/connector_sink_checkpoint_rollback.pony

```diff
@@ -0,0 +1,465 @@
+
+class _CpRbRollingBackResumed is _CpRbOps
+  let _debug: LogSevCat = Log.make_sev_cat(Log.debug(), Log.conn_sink())
+  let _barrier_token: CheckpointBarrierToken
+
+  fun name(): String => __loc.type_name()
+
+  new create(barrier_token: CheckpointBarrierToken) =>
+    _barrier_token = barrier_token
+
+  fun ref enter(sink: ConnectorSink ref) =>
+    sink.cprb_send_advertise_status(true)
+
+  fun ref conn_ready(sink: ConnectorSink ref) =>
+    _CpRbTransition(this, _CpRbWaitingForCheckpoint, sink)
+
+  fun ref abort_next_checkpoint(sink: ConnectorSink ref) =>
+    // The _ExtConnOps turned advertise_status off when this message was
+    // sent.  This state needs it on again.
+    sink.cprb_send_advertise_status(true)
+
+    // To avoid deadlock, resend rollback_info, just in case.
+    sink.cprb_send_rollback_info(_barrier_token)
+
+  fun ref cp_barrier_complete(sink: ConnectorSink ref,
+    barrier_token: CheckpointBarrierToken, queued: Array[SinkPhaseQueued])
+  =>
+    @l(Log.info(), Log.conn_sink(),
+      "Checkpoint %s arrived at _CpRbRollingBackResumed.cp_barrier_complete, aborting it".cstring(),
+        barrier_token.string().cstring())
+    let txn_id = sink.cprb_make_txn_id_string(barrier_token.id)
+    sink.cprb_send_abort_to_barrier_coordinator(barrier_token, txn_id)
+
+    // Ignore the queued arg: we are already queuing before entering
+    // this FSM state, so shear can't happen.
+    _CpRbTransition(this, _CpRbPreparedForRollback, sink)
+
+class _CpRbWaitingForCheckpoint is _CpRbOps
+  let _debug: LogSevCat = Log.make_sev_cat(Log.debug(), Log.conn_sink())
+  fun name(): String => __loc.type_name()
+
+  fun ref enter(sink: ConnectorSink ref) =>
+    sink.resume_processing_messages_queued()
+
+  fun ref abort_next_checkpoint(sink: ConnectorSink ref) =>
+    _CpRbTransition(this, _CpRbAbortCheckpoint(None), sink)
+
+  fun ref cp_barrier_complete(sink: ConnectorSink ref,
+    barrier_token: CheckpointBarrierToken, queued: Array[SinkPhaseQueued])
+  =>
+    _CpRbTransition(this, _CpRbCPStarts(barrier_token, queued), sink)
+
+  fun ref prepare_for_rollback(sink: ConnectorSink ref) =>
+    _CpRbTransition(this, _CpRbPreparedForRollback(where shear_risk=true), sink)
+
+  fun ref rollback(sink: ConnectorSink ref,
+    barrier_token: CheckpointBarrierToken)
+  =>
+    // We are very early in the startup process and are recovering.
+    // Let's roll back.
+    // However, we need to change phase first, and we must avoid shear.
+    @ll(_debug,
+      "Rollback: early rollback at %s!".cstring(), __loc.type_name().cstring())
+    sink.cprb_send_advertise_status(false)
+    _ChangeSinkPhaseQueueMsgsForwardTokens(sink where shear_risk = true)
+    _CpRbTransition(this, _CpRbRollingBack(barrier_token), sink)
+
+primitive _ChangeSinkPhaseQueueMsgsForwardTokens
+  fun apply(sink: ConnectorSink ref, shear_risk: Bool = false) =>
+    """
+    I call "shear risk" the phenomenon when this sink's SinkPhase object
```

Comment:
The only case where we should change phase in the middle of processing a barrier is if that barrier has been overridden, and in that case we need mechanisms to ensure we ignore remaining tokens. This should be guaranteed by ordering: i.e. an overriding barrier will always follow an active barrier that it overrides, on all channels. Once we receive a higher priority barrier on any channel, we should ignore a lower priority barrier on any other channel. If it's possible to change phase while we still expect 1 or more barriers, that sounds like a bug. When do you see this happen?

Reply:
Q: "When do you see this happen?"

A: Usually within 10 minutes of any `master-crasher.sh` run with any type of crash.

This PR changes the fundamental control the flow of app messages out of the sink, by pushing the job to the `SinkPhase`.  And every single one of the checkpoint/rollback component FSM transitions that *are not* caused by a barrier-complete event can (and definitely does) cause "shear" on a partially-observed barrier: prepare_for_rollback, rollback, abort_next_checkpoint, phase1_commit, phase1_abort, conn_ready.

Reply:
I think we need to discuss this in more depth. It feels wrong to me here, and I think it will be easy for future editors of this code to introduce bugs.